Add TestGrid Documentation and Linting (#10)
CSPI-QE/ocp-ci-docs: docs/OCP_CI_Tutorials/Containers/Container_Creation_Guide.md

    - [Execute in the Container](#execute-in-the-container)

## Introduction

> Prior to discussing the how to use containers in OpenShift CI, it is important to understand the basic structure of a CSPI test in OpenShift CI. 

Reviewer: ```suggestion
Prior to discussing how to use containers in OpenShift CI, it is important to understand the basic structure of a CSPI test in OpenShift CI. 
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

> Prior to discussing the how to use containers in OpenShift CI, it is important to understand the basic structure of a CSPI test in OpenShift CI. 

Reviewer 2: ```suggestion
Before we discuss how to use containers in OpenShift CI, it is important to understand the basic structure of a CSPI test in OpenShift CI. 
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

```mermaid
    graph TD;
        ocp[OpenShift CI]-->prow[Prow Cluster];
        prow[Prow Cluster]-->test[Test Cluster]
```

`Figure 1: _Cluster Graph`

> Figure 1 above shows a simple flow chart to demonstrate that there are really two clusters provisioned when you execute a CSPI job in OpenShift CI.

Reviewer 2: ```suggestion
Figure 1 above shows a simple flowchart which illustrates that there are really two clusters provisioned when you execute a CSPI job in OpenShift CI.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

1. **Prow Cluster**

>    - This cluster is the cluster that actually executes the various steps (chains, workflows, refs, _etc._) of a test.

Reviewer 2: ```suggestion
   - The Prow cluster actually executes the various steps (chains, workflows, refs, _etc._) of a test.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

>    - This cluster isn't provisioned new each time. There is a new project/namespace in this cluster provisioned and deprovisioned for every execution of OpenShift CI.

Reviewer 2: ```suggestion
   - The Prow cluster is not newly provisioned each time. There is a new project/namespace in this cluster that provisions and deprovisions for every execution of OpenShift CI.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

The containers created for a specific scenario should be able to run any interop test written for that scenario. When writing the tests for a CSPI scenario, try to think about how they will be executed inside of a container as well as how you will target a separate cluster.

### Creating Containerized Tests

> For the following example, a very simple Python test will be utilized. This is just to help wrap your head around the idea of testing within a container.

Reviewer 2: ```suggestion
The following example uses a very simple Python test. This example is used to help understand the basic concept of testing within a container.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

> > If your tests require any variables prior to executing the tests, please work with the CSPI team ahead of time which variables you need and (if needed) how to retrieve those variables. It is possible to get a variable (i.e. the web address of the test cluster) during a previous container's execution and that variable can be placed in a file in the `SHARED_DIR` for usage in other containers.

Reviewer: It appears that some words are missing from the following portion of this line: "please work with the CSPI team ahead of time which variables"

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

> > If your tests require any variables prior to executing the tests, please work with the CSPI team ahead of time which variables you need and (if needed) how to retrieve those variables. It is possible to get a variable (i.e. the web address of the test cluster) during a previous container's execution and that variable can be placed in a file in the `SHARED_DIR` for usage in other containers.

Reviewer: ```suggestion
> If your tests require any variables prior to executing the tests, please work with the CSPI team ahead of time which variables you need and (if needed) how to retrieve those variables. It is possible to get a variable (e.g. the web address of the test cluster) during a previous container's execution and that variable can be placed in a file in the `SHARED_DIR` for usage in other containers.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

> > If your tests require any variables prior to executing the tests, please work with the CSPI team ahead of time which variables you need and (if needed) how to retrieve those variables. It is possible to get a variable (i.e. the web address of the test cluster) during a previous container's execution and that variable can be placed in a file in the `SHARED_DIR` for usage in other containers.

Reviewer 2: ```suggestion
>Before executing your tests, please work with the CSPI team if your tests require any variables and (if needed) how to retrieve those variables. It is possible to get a variable (e.g. the web address of the test cluster) during a previous container's execution, and that variable can be placed in a file in the `SHARED_DIR` for usage in other containers.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

> > If your tests require any variables prior to executing the tests, please work with the CSPI team ahead of time which variables you need and (if needed) how to retrieve those variables. It is possible to get a variable (i.e. the web address of the test cluster) during a previous container's execution and that variable can be placed in a file in the `SHARED_DIR` for usage in other containers.

Author: ```suggestion
> If your tests require any variables prior to executing the tests, please tell the CSPI team which variables are needed and (if possible) how to retrieve those variables. It is possible to get a variable (e.g. the web address of the test cluster) during a previous container's execution and that variable can be placed in a file in the `SHARED_DIR` for usage in other containers.
```

#### The Container

Now that we have tests created, we can create a Dockerfile that will build an image to run the tests. Before writing the Dockerfile, it can be helpful to list any requirements to run the tests. To run the tests above, the container will need:

1. Python

> 2. pip

Reviewer: ```suggestion
2. Pip
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

```

### Executing Containerized Tests in OpenShift CI

Using a container is very easy in OpenShift CI. There are several ways it can be done (from a registry, a Dockerfile, an image stream, _etc._). However, for the purposes of this guide, we will focus on using the Dockerfile.

#### Define the Container

To define a container in OpenShift CI, two stanzas need to be added to the configuration file in OpenShift CI. Keep in mind, the configuration files in OpenShift CI are named based on the organization, repository, and branch of the test repository. So when we specify the Dockerfile's location or the context of the build, it is relative to the root of the testing repository. 

> Pretend we have a test repository in the `CSPI-QE` organization named `mock_tests` and we'd like to use the `main` branch of the repository to test with. The configuration file in the `openshift/release` repository would be `ci-operator/config/cspi-qe/mock_tests/cspi-qe-mock_tests-main.yaml`. 

Reviewer: Since the organization name is case sensitive then we should use CSPI-QE instead of cspi-qe

Author: I wouldn't mind that change, but we'd have to bring that up with Adam to change the org name in GitHub.

Author: ```suggestion
Pretend we have a test repository in the `CSPI-QE` organization named `mock_tests` and we'd like to use the `main` branch of the repository to test with. The configuration file in the `openshift/release` repository would be `ci-operator/config/CSPI-QE/mock_tests/CSPI-QE-mock_tests-main.yaml`. 
```
@madunn is this what you are looking for? I'm not sure if the file name is also case-sensitive.

Inside of this pretend repository there is a directory named `dockerfiles` containing the Dockerfile we made above. In the root of the repository in the test file, `test_mock_tests.py`.

> The following is an abbreviated version of the configuration file for our test repository. The most important stanza to look at for this guide are:

Reviewer: ```suggestion
The following is an abbreviated version of the configuration file for our test repository. The most important stanzas to look at for this guide are:
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)
